File: src/main/java/org/dukcode/ps/codetree/trail02/chapter06/lesson01/ChallengeBestPlaceOf132.java
package org.dukcode.ps.codetree.trail02.chapter06.lesson01;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/challenge-best-place-of-13-2/description">LINK</a>
 */
public class ChallengeBestPlaceOf132 {

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  private static int n;
  private static int[][] board;
  private static int[][] taken;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    n = Integer.parseInt(br.readLine());
    board = new int[n][n];
    for (int y = 0; y < n; y++) {
      st = new StringTokenizer(br.readLine());
      for (int x = 0; x < n; x++) {
        board[y][x] = Integer.parseInt(st.nextToken());
      }
    }

    taken = new int[n][n];

    bw.write(String.valueOf(solve(0, 0)));

    br.close();
    bw.close();

  }

  private static int solve(int cntDraw, int cnt) {
    if (cntDraw == 2) {
      return cnt;
    }

    int ret = 0;
    for (int y = 0; y < n; y++) {
      for (int x = 0; x < n; x++) {
        int cntBlock = draw(y, x, 1);

        if (cntBlock != -1) {
          ret = Math.max(ret, solve(cntDraw + 1, cnt + cntBlock));
        }

        draw(y, x, -1);
      }
    }

    return ret;
  }

  private static int draw(int y, int x, int delta) {
    int cnt = 0;
    boolean possible = true;
    for (int i = 0; i < 3; ++i) {
      int nx = x + i;

      if (nx >= n) {
        possible = false;
        continue;
      }

      if (taken[y][nx] != 0) {
        possible = false;
      }
      cnt += board[y][nx];
      taken[y][nx] += delta;

    }

    return possible ? cnt : -1;
  }

}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter06/lesson01/ChallengeBestPlaceOf132.kt
package org.dukcode.ps.codetree.trail02.chapter06.lesson01.challengeBestPlaceOf132

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer
import kotlin.math.abs

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/challenge-best-place-of-13-2/description">LINK</a>
 */
fun main() {
    val n = br.readLine().toInt()
    val board = Array(n) { parseInts() }

    val ans =
        (0 until n).maxOf { y1 ->
            (0 until n - 2).maxOf { x1 ->
                (0 until n).maxOf { y2 ->
                    (0 until n - 2).maxOf { x2 ->
                        if (isOverlapped(y1, x1, y2, x2)) {
                            -1
                        } else {
                            blockSum(board, y1, x1) + blockSum(board, y2, x2)
                        }
                    }
                }
            }
        }

    bw.write(ans.toString())

    bw.flush()
}

private fun blockSum(
    board: Array<IntArray>,
    y: Int,
    x: Int,
): Int = (0 until 3).sumOf { dx -> board[y][x + dx] }

private fun isOverlapped(
    y1: Int,
    x1: Int,
    y2: Int,
    x2: Int,
): Boolean = y1 == y2 && abs(x1 - x2) < 3

private fun parseInts() =
    StringTokenizer(br.readLine()).run {
        IntArray(countTokens()) {
            nextToken().toInt()
        }
    }
